### DISCUSSION TITLE
import requests issue

### DISCUSSION THREAD

**Discussion by cataYT** [modular/modular#828] [Q&A] [answered]
from python import Python
import requests

fn main():

    # URL you want to send a GET request to
    let url = 'https://jsonplaceholder.typicode.com/todos/1'

    # Send the GET request
    let response = requests.delete(url)

    # Check if the request was successful (status code 200)
    if response.status_code == 200:
        print('Request was successful')
        print('Response content:')
        print(response.text)
    else:
        print(f'Request failed with status code {response.status_code}')

    let save = str(input("Do you want to save the response?: "))

    if save.lower() == "yes" or save.lower() == "y":

        let saveFile = open("response.json", "w")

        saveFile.write(f"{response.text}")

        saveFile.close()
        
In the above program, it says unable to locate requests so my question is there a mojo version for this?

**Comment by MikeCase** [comment, marked as answer]
There's not a Mojo version of requests yet, but I believe you can try another way. 

```py
from python import Python

fn main():
    try:
        let requests = Python.import_module("requests")
        # URL you want to send a GET request to
        let url = 'https://jsonplaceholder.typicode.com/todos/1'

        # Send the GET request
        let response = requests.delete(url)

        # Check if the request was successful (status code 200)
        if response.status_code == 200:
            print('Request was successful')
            print('Response content:')
            print(response.text)
        else:
            print('Request failed with status code', response.status_code)

    except:
        print("didn't work")
```

I don't yet see a way to write to files built in, you may have to code that by hand. Unfortunately that's above my level. But I'm trying to learn everything I can about the language while it's new and minimal so over time I can just add to my knowledge.

**Reply by cataYT** [reply]
this one worked! thanks

**Reply by MikeCase** [reply]
I know this has been answered already, and that I provided the correct answer... but I happened to have this problem opened in my editor when I sat down tonight. 
So I've been messing around with it. I've come up with this solution to provide both requests usage and write file functionality.


```py
from python import Python


fn write_file(content: String) raises:
    '''
    Uses pythons builtins to open and write a file.
    '''
    let builtins = Python.import_module("builtins")
    let save = builtins.str(builtins.input("Save a file?"))
    if save.lower() == "yes" or save.lower() == "y":
        let saveFile = builtins.open("test.json", "a")
        _ = saveFile.write(content)
        _ = saveFile.close()


fn main() raises:
    '''
    Makes a request with pythons request module and returns the response.
    '''

    let requests = Python.import_module("requests")
    # URL you want to send a GET request to
    let url = 'https://jsonplaceholder.typicode.com/todos/2'

    # Send the GET request
    let response = requests.get(url)

    # Check if the request was successful (status code 200)
    if response.status_code == 200:
        print('Request was successful')
        print('Response content:')
        print(response.text)
        let resp = response.text
        write_file(resp.to_string())
    else:
        print('Request failed with status code', response.status_code)
```

**Comment by rarebreed** [comment]
In mojo, you can't just import python modules like in python.  So this won't work

```python
import requests
```

And MikeCase is also right that there doesn't seem to be a way in mojo yet to do any kind of mojo native File or Networking IO.  You can get the `open` function through python's `__builtins__` (which normally is automatically imported in python, but you can explicitly import it too with `import builtins`).  Also you won't be able to use a Context Manager in mojo yet.

I haven't tested this myself, but try this

```python
fn main():
    let bi = Python.import_module("builtins")
    let requests = Python.import_module("requests")

    let url = 'https://jsonplaceholder.typicode.com/todos/1'
  
    # Send the GET request
    let response = requests.delete(url)
  
    # Check if the request was successful (status code 200)
    if response.status_code == 200:
        print('Request was successful')
        print('Response content:')
        print(response.text)
    else:
        print(f'Request failed with status code {response.status_code}')
  
    let save = bi.str(bi.input("Do you want to save the response?: "))
    if save.lower() == "yes" or save.lower() == "y":
        let saveFile = bi.open("response.json", "w")
        saveFile.write(f"{response.text}")
        saveFile.close()
```

You may also get a warning about a shared python library not being found if you compile it as an executable

> Python interoperability might fail when running a compiled Mojo program, with the message Unable to locate a suitable libpython, please set MOJO_PYTHON_LIBRARY. This is because we currently do not embed the Python version into the Mojo binary. For details and the workaround, see [issue #551](https://github.com/modularml/mojo/issues/551).

**Reply by cataYT** [reply]
it gives me this error:

./test.mojo:2:14: error: use of unknown declaration 'Python', 'fn' declarations require explicit variable declarations
    let bi = Python.import_module("builtins")
             ^~~~~~
./test.mojo:3:20: error: use of unknown declaration 'Python', 'fn' declarations require explicit variable declarations
    let requests = Python.import_module("requests")
                   ^~~~~~
mojo: error: failed to parse the provided Mojo

**Reply by rarebreed** [reply]
Sorry, I didn't include the full mojo file.  At the top you need to import Python

```python
from python import Python

fn main():
    let bi = Python.import_module("builtins")
    let requests = Python.import_module("requests")

    let url = 'https://jsonplaceholder.typicode.com/todos/1'
  
    # Send the GET request
    let response = requests.delete(url)
  
    # Check if the request was successful (status code 200)
    if response.status_code == 200:
        print('Request was successful')
        print('Response content:')
        print(response.text)
    else:
        print(f'Request failed with status code {response.status_code}')
  
    let save = bi.str(bi.input("Do you want to save the response?: "))
    if save.lower() == "yes" or save.lower() == "y":
        let saveFile = bi.open("response.json", "w")
        saveFile.write(f"{response.text}")
        saveFile.close()
```
